0000-classical-stores.md

```diff
@@ -0,0 +1,493 @@
+
+Out of scope:
+
+This RFC does not propose making `Expr` nodes available in any more places than they already are, nor adding any new types to the system.
+In particular, this RFC does not yet address the question of making `Expr` nodes usable in `params` locations (such as the angle inputs to gates).
+
+
+## Motivation and Benefit
+
+Qiskit needs a way to represent the sorts of "parametric update" that we expect hardware vendors to be implementing in the near future.
+For example, we will ideally be able to use a single compiled circuit (both in the Qiskit and hardware senses of the term "compiled") that accepts certain parameters as inputs at runtime, which may affect conditional expressions or gate angles.
+This RFC only addresses the question of defining new classical-variable storage locations and writing to them for Qiskit within its existing typed `Expr` system, which is currently limited to unsigned integers and Booleans and cannot be used for gate angles.
+
+For example, Pauli-twirling workloads want to do something like
+```python
+from qiskit.circuit import QuantumCircuit
+from qiskit.circuit.classical import expr
+
+qc = QuantumCircuit(2)
+with qc.if_test(expr.value(TWIRL_BITS_0)):
+    qc.x(0)
+with qc.if_test(expr.value(TWIRL_BITS_1)):
+    qc.x(1)
+```
+where `TWIRL_BITS_*` are some values defined by the particular set of twirls to be done, and are known before the circuit begins execution, but not _necessarily_ at the time of the definition of the circuit, or the circuit should be run with many different values for them.
+These may also want to be compared to the results of mid-circuit measurements, or have arbitrary run-time classical expressions built from them.
+Right now, Qiskit provides no way to parametrise `if` conditions in any form; even the compile-time rewriting of `assign_parameters` will not support them, because `ParameterExpression` objects are (deliberately) not valid in `condition` fields.
+
+Similarly, a user may need to calculate a value based on some measured bits, potentially dependent on some `if` condition, and store the value in classical memory for later use in the circuit.
+In other words, they need variable assignment at runtime.
+
+
+## Design Proposal
+
+At a high level, this proposal concerns adding concepts equivalent to the OpenQASM 3 `input` (and potentially `output`) and variable-assignment-by-expression statements to Qiskit, within the classical types we already support.
+I do not yet propose removing `Clbit` or `ClassicalRegister`, nor trying to change them into some other form; for the time being, they retain their "special-case" status.
+
+The additional user interface for defining and working with variables on `QuantumCircuit` is:
+
+```python
+class QuantumCircuit:
+    def __init__(
+        self,
+        ...,  # existing structure
+        *,
+        inputs: Iterable[expr.Var] = (),
+        captures: Iterable[expr.Var] = (),
```

Comment:
Are you at all concerned about the conflict with the other way "capture" is used in openpulse? https://openqasm.com/language/openpulse.html#capture-instruction

Reply:
I see below, though, that you are using this exactly in the sense of C++ "captures" for closures. Well, it is an unfortunate word collision.

Reply:
More precisely, a "closure" is a structure (the function plus an environment) that may have "captures" (the variables from the environment), though we often are a bit lax in terminology and use "closure" to refer to both. I was trying to use the proper name here.